websites/Y/YouTube Music/presence.ts

```diff
@@ -2,90 +2,263 @@ const presence = new Presence({
 });
 
-let prevMetadata: {
-	title: string;
-	startedAt: number;
-};
+let prevTitleAuthor = "",
+	presenceData: PresenceData,
+	mediaTimestamps: [number, number],
+	oldPath: string,
+	startTimestamp: number,
+	videoListenerAttached = false;
 
 presence.on("UpdateData", async () => {
-	const title = navigator.mediaSession.metadata?.title,
-		video = document.querySelector<HTMLVideoElement>(".video-stream"),
-		progressBar = document.querySelector<HTMLElement>("#progress-bar"),
-		repeatMode = document
-			.querySelector('ytmusic-player-bar[slot="player-bar"]')
-			.getAttribute("repeat-Mode_"),
-		[buttons, timestamps, cover] = await Promise.all([
+	const [
+			showButtons,
+			showTimestamps,
+			showCover,
+			hidePaused,
+			showBrowsing,
+			privacyMode,
+		] = await Promise.all([
 			presence.getSetting<boolean>("buttons"),
 			presence.getSetting<boolean>("timestamps"),
 			presence.getSetting<boolean>("cover"),
-		]);
-	if (title && !isNaN(video.duration)) {
-		const endTimestamp =
-				Date.now() / 1000 +
-				Number(progressBar.getAttribute("aria-valuemax")) -
-				Number(progressBar.getAttribute("value")),
-			watchID = document
-				.querySelector<HTMLAnchorElement>("a.ytp-title-link.yt-uix-sessionlink")
-				.href.match(/v=([^&#]{5,})/)[1],
-			presenceData: PresenceData = {
-				details: title,
-				state: [
-					navigator.mediaSession.metadata.artist,
-					navigator.mediaSession.metadata.album,
-				]
-					.filter(Boolean)
-					.join(" - "),
-				largeImageKey: cover
-					? navigator.mediaSession.metadata.artwork[0].src
-					: "ytm_lg",
-				smallImageKey: video.paused
+			presence.getSetting<boolean>("hidePaused"),
+			presence.getSetting<boolean>("browsing"),
+			presence.getSetting<boolean>("privacy"),
+		]),
+		{ mediaSession } = navigator,
+		watchID = document
+			.querySelector<HTMLAnchorElement>("a.ytp-title-link.yt-uix-sessionlink")
+			.href.match(/v=([^&#]{5,})/)?.[1],
+		repeatMode = document
+			.querySelector('ytmusic-player-bar[slot="player-bar"]')
+			.getAttribute("repeat-Mode_"),
+		videoElement =
+			document.querySelector<HTMLVideoElement>("video.video-stream");
+
+	if (videoElement) {
+		if (!videoListenerAttached) {
+			const videoListener = () => {
+				mediaTimestamps = presence.getTimestampsfromMedia(videoElement);
+				//* Don't ask me why the above function doesn't floor the end timestamp
+				mediaTimestamps[1] = Math.floor(mediaTimestamps[1]);
+			};
+
+			//* If video scrobbled, update timestamps
+			videoElement.addEventListener("seeked", videoListener);
+			//* If video resumes playing, update timestamps
+			videoElement.addEventListener("play", videoListener);
+
+			videoListenerAttached = true;
+		}
+		//* Element got removed from the DOM (eg, song with song/video switch)
+	} else {
+		prevTitleAuthor = "";
+		videoListenerAttached = false;
+	}
+
+	presenceData = null;
+
+	if (hidePaused && mediaSession.playbackState !== "playing")
+		return presence.clearActivity();
```

Comment:
This function is broken

Reply:
If it works for your small tests, it doesn't neccessarily mean it will work all the time. I have seen countless instances of the method not working

Reply:
It's still good practice to have it tho, as it's going to be fixed in v3 nontheless

Reply:
This doesn't make sense

Reply:
clearActivity should work, if it doesn't then thats a bug, that doesn't mean it shouldn't be used at all.